yodeploy/repository.py

```diff
@@ -97,11 +102,29 @@ def get_metadata(self, path):
         with open(meta_fn) as f:
             return json.load(f)
 
-    def put(self, path, fp, metadata=None):
-        '''
-        Store a file (fp).
+    def put(self, path, data, metadata=None):
+        """
+        Store a file (fp), stream, unicode string, or byte string.
```

Comment:
It does, it does. While isolated, it's admittedly still gross. We open files in a lot of places and then `put()` them into the repository. In order to remove the unicode support that I've added here, I'd have to make all of those opens explicitly byte streams so that the behavior is consistent in 2/3 even though it should really be safe either way in many cases (the places it isn't, I did make explicit byte streams). That would additionally mean I have to make certain strings explicitly byte strings in a handful of places where we're writing generated data to the repository.

I originally feared that such changes would lead to python 3 specific coercion logic in a number of places, so my approach was to instead consolidate handling the various types of file streams and strings here and to cover the additional behaviors with unit tests. The result was that Python 2 continues to do things with byte strings/streams and behaves like it always did wheras Python 3 takes a unicode (almost) everywhere approach and these changes make that just work. This allows us to incrementally change things over to unicode everywhere possible under python 2 or just leave things as they are, which should also be safe for our purposes.

I originally felt more strongly about this than I do now, so it's kind've a toss up for me at this point. Behavioral consistency is nice, but on the other hand I think I would expect `put()` to "just work" if I was messing with a file handle while working on some Python 3 implementation.